--- a/board/omap3beagle/omap3beagle.c	2011-04-08 07:08:24.000000000 -0700
+++ b/board/omap3beagle/omap3beagle.c	2011-10-27 02:46:46.000000000 -0700
@@ -48,6 +48,7 @@
 #define REVISION_CX		0x6
 #define REVISION_C4		0x5
 #define REVISION_XM		0x0
+#define REVISION_XM_C		0x2
 
 /* Used to index into DPLL parameter tables */
 struct dpll_param {
@@ -322,8 +323,8 @@
 	/* setup sdrc to ball mux */
 	__raw_writel(SDP_SDRC_SHARING, SDRC_SHARING);
 
-	if (beagle_revision() == REVISION_XM) {
-		__raw_writel(0x2, SDRC_CS_CFG); /* 256MB/bank */
+	if (beagle_revision() == REVISION_XM_C) {
+		__raw_writel(0x3, SDRC_CS_CFG); /* 256MB/bank */
 		__raw_writel(SDP_SDRC_MDCFG_0_DDR_XM, SDRC_MCFG_0);
 		__raw_writel(SDP_SDRC_MDCFG_0_DDR_XM, SDRC_MCFG_1);
 		__raw_writel(MICRON_V_ACTIMA_200, SDRC_ACTIM_CTRLA_0);
@@ -894,6 +895,9 @@
 	case REVISION_XM:
 		printf("Beagle xM Rev A\n");
 		break;
+	case REVISION_XM_C:
+		printf("Beagle xM Rev C\n");
+		break;
 	default:
 		printf("Beagle unknown 0x%02x\n", rev);
 	}
@@ -1200,8 +1204,10 @@
 
 		if (nand_chip()) {
 #ifdef CFG_PRINTF
+#ifndef REVISION_XM_C
 			printf("Unsupported Chip!\n");
 #endif
+#endif
 			return 0;
 		}
 
